### DISCUSSION TITLE
Make-Whole Tables?

### DISCUSSION THREAD

**Discussion by gfarah19** [domokane/FinancePy#212] [Q&A]
hi @domokane , what you've built is amazing! I had a question regarding convertible bonds... is there functionality built in for valuing a convertible bond that has a make-whole table (whether it's additional stock or capital granted during a trigger event)? If not, do you know of any libraries that could hook into FinancePy to calculate greeks/value given the make-whole table? Thank you!

**Comment by heykav** [comment]
Checked the source (`financepy/products/bonds/bond_convertible.py`) to be sure rather than guess: it's not in there. `ConvertibleBond` supports a flat call-price schedule (`call_dts`/`call_prices` — one price per call date) but there's no make-whole table concept, i.e. no mechanism for a stock-price-dependent additional-shares/compensation grant triggered on early call. Those are genuinely different structures — a call price schedule is one number per date; a make-whole table is a 2D grid (call date × trigger stock price) — and the model here only implements the former.

So no, this isn't something you can drive through FinancePy's existing `ConvertibleBond` as-is. If it's a hard requirement, the honest options are: (1) fork/extend `bond_convertible.py` and its pricing tree (`bond_convertible_bs_tree.py`) to interpolate a make-whole table into the exercise boundary at each call date, which is a real but bounded piece of work since the tree structure is already there, or (2) look at a purpose-built CB pricer that already models make-whole provisions (this tends to be a Bloomberg CB<GO>-style feature, so vendor tools that replicate that methodology are more likely to have it out of the box than general-purpose open-source libraries).

**Comment by domokane** [comment]
Hi! If you wish to consider either extending the current tree or making a separate tree let me know in advance. I am doing other work right now and will not get to convertible bonds for a few months. If you don't do it I might look at doing it then. Or raise it as an issue someone else can do.

**Comment by heykav** [comment]
Thanks — happy to take this on, and I'd rather do it properly than fast, so no pressure on your end either way.

My inclination is to extend the existing tree (`bond_convertible_bs_tree.py`) rather than build a separate one: the backward-induction structure and exercise-boundary logic are already there, and a make-whole table is really just another input to interpolate into that boundary at each call date rather than a different pricing methodology. A separate tree would only make sense if make-whole CBs turn out to need a genuinely different state space, and I don't think they do — but I'll confirm that once I'm actually inside the code rather than assuming from the outside.

A couple of things I want to get right before writing anything, since getting the table semantics wrong would be worse than not having the feature at all:

1. **Table shape** — my working assumption is a 2D grid (call date × trigger stock price) returning either an additional-shares figure or a make-whole premium, which is the general shape these tables take in real CB indentures. Is that matched to what you had in mind, or is there a specific convention/format you'd want it to accept?
2. **Interpolation between grid points** — usual practice is bilinear interpolation across both date and price axes, with flat extrapolation outside the table's range. Want to confirm that's the right default before I bake it in.

I'll open this as a proper issue to track the design before any PR, so there's a paper trail of the approach rather than it just showing up as a diff. No firm timeline from me either — I'd rather get the methodology right than rush it, especially since this isn't something with an easy source-of-truth test the way a lot of the rest of the library seems to have.

**Comment by domokane** [comment]
Hi - OK. Is there a bloomberg model or other model or paper that you can use as a reference ? As I said I am busy on other library aspects so I am happy if you complete this and submit a PR. No rush.
